### PR TITLE
feat: add validators to check values in services

### PR DESCRIPTION
**Pull request by luka-nextcloud**
Signed-off-by: Luka Trovic <luka@nextcloud.com>

* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [ ] Code is properly formatted
- [ ] Sign-off message is added to all commits
- [ ] Tests (unit, integration, api and/or acceptance) are included
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by stefan-niedermann** [comment]
Any breaking changes in here? I didn't see one at the first look, but just to make sure 😉

File: lib/Validators/BaseValidator.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	 */
+	public function check(array $data)
+	{
+		return $this->validate($data);
+	}
+
+	// TODO: check length base on the database field length
+	private function checkLength ($value) {
+		if (!is_string($value)) return true;
+		return strlen($value) <= self::MAX_LENGTH;
```
Comment (nickvergessen) [inline lib/Validators/BaseValidator.php:71]: might always want to compare `mb_strlen` here, to not restrict languages like russian too much.
```php
var_dump(strlen('ипсум')); // int(10)
var_dump(mb_strlen('ипсум')); // int(5)
```

When the DB is configured with utf8mb4 support it is also only 5 chars on the db length.

Comment (juliushaertl) [reply lib/Validators/BaseValidator.php:71]: This method seems to be unused anyways and mb_strlen is used in the getSize method, so I think you could just drop that.

File: lib/Validators/BaseValidator.php [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	/**
+	 * @param array $data
+	 * @return bool
+	 */
+	public function check(array $data)
+	{
+		return $this->validate($data);
+	}
```
Comment (juliushaertl) [inline lib/Validators/BaseValidator.php:66]:
```suggestion
	public function check(array $data): bool {
		return $this->validate($data);
	}
```

File: lib/Validators/AssignmentServiceValidator.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?php
+declare(strict_types=1);
+
```
Comment (juliushaertl) [inline lib/Validators/AssignmentServiceValidator.php:3]: Copyright header is missing ;)

File: lib/Service/StackService.php [hunk 4]
```diff
@@ -291,22 +285,12 @@ public function delete($id) {
 	 * @throws BadRequestException
 	 */
 	public function update($id, $title, $boardId, $order, $deletedAt) {
+		$this->stackServiceValidator->check(compact('title', 'boardId', 'order'));
+
 		if (is_numeric($id) === false) {
```
Comment (juliushaertl) [inline lib/Service/StackService.php:290]: Can we also validate the id through the validator?

File: lib/Service/LabelService.php [hunk 5]
```diff
@@ -144,17 +144,12 @@ public function delete($id) {
 	 * @throws BadRequestException
 	 */
 	public function update($id, $title, $color) {
+		$this->labelServiceValidator->check(compact('title', 'color', 'boardId'));
+
 		if (is_numeric($id) === false) {
```
Comment (juliushaertl) [inline lib/Service/LabelService.php:149]: Can we also validate the id through the validator here?

File: lib/Validators/BaseValidator.php (hunk 2 above)
Comment (juliushaertl) [reply lib/Validators/BaseValidator.php:66]: Actually it seems to throw so maybe rather use void and annotate the exception that is being thrown here.

File: lib/Validators/BaseValidator.php [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+	// TODO: check length base on the database field length
+	private function checkLength ($value) {
+		if (!is_string($value)) return true;
+		return strlen($value) <= self::MAX_LENGTH;
+	}
+
+	private function numeric ($value) {
```
Comment (juliushaertl) [inline lib/Validators/BaseValidator.php:74]:
```suggestion
	private function numeric ($value): bool {
```

File: lib/Validators/BaseValidator.php [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		return is_numeric($value);
+	}
+
+	private function bool ($value) {
```
Comment (juliushaertl) [inline lib/Validators/BaseValidator.php:78]:
```suggestion
	private function bool ($value): bool {
```

File: lib/Validators/BaseValidator.php [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+		return is_bool($value);
+	}
+
+	private function not_false ($value) {
```
Comment (juliushaertl) [inline lib/Validators/BaseValidator.php:82]:
```suggestion
	private function not_false ($value): bool {
```

File: lib/Validators/BaseValidator.php [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+		return ($value !== false) && ($value !== 'false');
+	}
+
+	private function not_null ($value) {
```
Comment (juliushaertl) [inline lib/Validators/BaseValidator.php:86]:
```suggestion
	private function not_null ($value): bool {
```

File: lib/Validators/BaseValidator.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		return !is_null($value);
+	}
+
+	private function not_empty ($value) {
```
Comment (juliushaertl) [inline lib/Validators/BaseValidator.php:90]:
```suggestion
	private function not_empty ($value): bool {
```

File: lib/Validators/BaseValidator.php [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	/**
+     * Get the size of an attribute.
+     *
+     * @param  mixed  $value
+     * @return mixed
+     */
+    protected function getSize($value)
+    {
```
Comment (juliushaertl) [inline lib/Validators/BaseValidator.php:114]:
```suggestion
	/**
     * Get the size of an attribute.
     *
     * @param  mixed  $value
     */
    protected function getSize($value): int {
```

**Comment by juliushaertl** [comment]
@luka-nextcloud Indentation seems off quite a bit in some places. Maybe double check your editor settings to use tab indentation. php-cs-fixer should however be able to auto fix those with `composer run cs:fix`

**Comment by luka-nextcloud** [comment]
@juliushaertl Please check again.

File: lib/Validators/BoardServiceValidator.php [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+declare(strict_types=1);
+
+namespace OCA\Deck\Validators;
+
+class BoardServiceValidator extends BaseValidator {
+	public function rules() {
+		return [
+			'id' => ['numeric'],
+			'mapper' => ['not_empty', 'not_null', 'not_false'],
+			'title' => ['not_empty', 'not_null', 'not_false', 'max:100', 'min:5'],
```
Comment (juliushaertl) [inline lib/Validators/BoardServiceValidator.php:34]:
```suggestion
			'title' => ['not_empty', 'not_null', 'not_false', 'max:100'],
```

This is causing the test failures it seems as it is a new restriction

File: lib/Validators/BoardServiceValidator.php [hunk 13]
```diff
@@ -0,0 +1,44 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016 Julius Härtl <jus@bitgrid.net>
+ *
+ * @author Julius Härtl <jus@bitgrid.net>
+ * @author Maxence Lange <maxence@artificial-owl.com>
```
Comment (juliushaertl) [inline lib/Validators/BoardServiceValidator.php:6]: Make sure to add you to the copyright, as you created that file :)

**Review by juliushaertl — APPROVED** [review]

**Comment by juliushaertl** [comment]
- stable23 https://github.com/nextcloud/deck/pull/4175
- stable24 https://github.com/nextcloud/deck/pull/4176

**Pull request merged**